File: valid_palindrome/ValidPalinadrome.java
class ValidPalinadrome {
    public boolean isPalindrome(String s) {
        if (s == null || s.length() < 2) {
            return true;
        }
        
        int left = 0;
        int right = s.length() - 1;
        while (left <= right) {
            char lch = s.charAt(left);
            char rch = s.charAt(right);
            if (!Character.isLetterOrDigit(lch)) {
                left++;
                continue;
            }
            if (!Character.isLetterOrDigit(rch)) {
                right--;
                continue;
            }
            
            if (lch != rch && (Character.isDigit(lch) || Character.isDigit(rch))) {
                return false;
            }
            if (lch - 'A' > 26) {
                lch -= 32;
            }
            if (rch - 'A' > 26) {
                rch -= 32;
            }
            if (lch != rch) {
                return false;
            }
            left++;
            right--;
        }
        
        return true;
    }
}

File: valid_palindrome/Valid_Palinadrome.cpp
class Valid_Palinadrome {
public:
    bool isPalindrome(string s) {
        if (s.empty() || s.size() < 2) {
            return true;
        }
        
        int left = 0;
        int right = s.size() - 1;
        while (left <= right) {
            if (isalpha(s[left]) == 0 && isdigit(s[left]) == 0) {
                left++;
                continue;
            }
            if (isalpha(s[right]) == 0 && isdigit(s[right]) == 0) {
                right--;
                continue;
            }
            if (s[left] != s[right] && (isdigit(s[left]) == 1 || isdigit(s[right]) == 1)) {
                return false;
            }
            
            if (s[left] - 'A' > 26) {
                s[left] -= 32;
            }
            if (s[right] - 'A' > 26) {
                s[right] -= 32;
            }
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        
        return true;
    }
};